sys/net/gnrc/doc.txt

```diff
@@ -0,0 +1,172 @@
+ * it is crucial for a new module that introduces a new thread to follow a certain programming
+ * construct if it desires to interact with other threads without blocking the system:
+ * Utilizing an `event loop`.
+ *
+ * Hence, a thread for @ref net_gnrc "GNRC" will usually consist of four basic steps.
+ * -# Initialize a message queue
+ * -# register for a @ref net_gnrc_nettype
+ * -# wait for a message
+ * -# react appropriately to a message and return to 3.
+ * @code{c}
+ * static msg_t _msg_q[Q_SZ];
+ * msg_init_queue(_msg_q, Q_SZ);
+ * gnrc_netreg_entry me_reg = { .demux_ctx = GNRC_NETREG_DEMUX_CTX_ALL, .pid = thread_getpid() };
+ * gnrc_netreg_register(GNRC_NETTYPE_IPV6, &me_reg);
+ * while (1) {
+ *     msg_receive(&msg);
+ *     switch (msg.type) {
+ *         case TYPE1:
+ *             callback1();
+ *             break;
+ *         ...
+ *     }
+ * }
+ * @endcode
+ *
+ * @note As an example have a look at the event loops of @ref net_gnrc_ipv6 and @ref net_gnrc_rpl
+ *
+ * Receiving / Transmitting Packets
+ * ================================
+ *
+ * Packets can be received or transmitted by interacting with the @ref net_gnrc_netapi.
+ *
+ * Receiving Packets
+ * -----------------
+ *
+ * The reception of a @ref net_gnrc_pkt from another thread is handled by the @ref net_gnrc_netapi
+ * module. In order to receive a @ref net_gnrc_pkt of a specific type, it is necessary to register
+ * for the appropriate @ref net_gnrc_nettype first. Your thread will then be able to receive
+ * certain commands defined in the @ref net_gnrc_netapi module
+ * (e.g. @ref net_gnrc_netapi::GNRC_NETAPI_MSG_TYPE_RCV) for all @ref net_gnrc_pkt "Packets"
+ * that your thread registered for.
+ *
+ * The following example will sketch how to receive incoming and outgoing UDP traffic on port 80.
+ * @code{c}
+ * static msg_t _msg_q[Q_SZ];
+ * msg_t msg, reply;
+ * reply.type = GNRC_NETAPI_MSG_TYPE_ACK;
+ * reply.content.value = -ENOTSUP;
+ * msg_init_queue(_msg_q, Q_SZ);
+ * gnrc_pktsnip_t *pkt = NULL;
+ * gnrc_netreg_entry me_reg = { .demux_ctx = 80, .pid = thread_getpid() };
+ * gnrc_netreg_register(GNRC_NETTYPE_UDP, &me_reg);
+ *
+ * while (1) {
+ *     msg_receive(&msg);
+ *     switch (msg.type) {
+ *         case GNRC_NETAPI_MSG_TYPE_RCV:
+ *             pkt = (gnrc_pktsnip_t *) msg.content.ptr;
+ *             _handle_incoming_pkt(pkt);
+ *             break;
+ *         case GNRC_NETAPI_MSG_TYPE_SND:
+ *             pkt = (gnrc_pktsnip_t *) msg.content.ptr;
+ *             _handle_outgoing_pkt(pkt);
+ *             break;
+ *          case GNRC_NETAPI_MSG_TYPE_SET:
+ *          case GNRC_NETAPI_MSG_TYPE_GET:
+ *             msg_reply(&msg, &reply);
+ *             break;
+ *         default:
+ *             break;
+ *     }
+ * }
+ * @endcode
+ * @note It is necessary to react with a reply message to the other thread for
+ *       @ref GNRC_NETAPI_MSG_TYPE_SET and @ref GNRC_NETAPI_MSG_TYPE_GET commands
```

Comment:
Should we mention here too, that it is necessary to unregister `me_reg` again, if you leave the function context?